src/ripple/websocket/Server.h

```diff
@@ -17,57 +17,57 @@
 */
 //==============================================================================
 
-#ifndef RIPPLED_RIPPLE_WEBSOCKET_WSDOORBASE_H
-#define RIPPLED_RIPPLE_WEBSOCKET_WSDOORBASE_H
+#ifndef RIPPLED_RIPPLE_WEBSOCKET_SERVER_H
+#define RIPPLED_RIPPLE_WEBSOCKET_SERVER_H
 
 #include <ripple/basics/Log.h>
 #include <ripple/websocket/WebSocket.h>
-#include <beast/cxx14/memory.h> // <memory>
 #include <beast/threads/Thread.h>
+#include <thread>
 
 namespace ripple {
 namespace websocket {
 
 template <class WebSocket>
-class Server
-    : public beast::Stoppable
-    , protected beast::Thread
+class Server : public beast::Stoppable
 {
 private:
-    // TODO: why is this recursive?
-    using LockType = std::recursive_mutex;
-    using ScopedLockType = std::lock_guard <LockType>;
-
     ServerDescription desc_;
-    LockType m_endpointLock;
+    std::recursive_mutex endpointMutex_;  // TODO: why is this recursive?
+    std::thread thread_;
     beast::Journal j_;
-    typename WebSocket::EndpointPtr m_endpoint;
+    typename WebSocket::EndpointPtr endpoint_;
 
 public:
     Server (ServerDescription const& desc)
-       : beast::Stoppable (WebSocket::versionName(), desc.source)
-        , Thread ("websocket")
+        : beast::Stoppable (WebSocket::versionName(), desc.source)
         , desc_(desc)
-       , j_ (desc.app.journal ("WebSocket"))
+        , j_ (desc.app.journal ("WebSocket"))
     {
-        startThread ();
+        thread_ = std::thread {&Server<WebSocket>::run, this};
     }
 
     ~Server ()
     {
-        stopThread ();
+        if (thread_.joinable())
+        {
+            doStop ();
+            thread_.join();
+        }
     }
 
 private:
-    void run () override
+    void run ()
     {
+        beast::Thread::setCurrentThreadName ("WebSocket");
+
         JLOG (j_.warning)
             << "Websocket: creating endpoint " << desc_.port;
 
         auto handler = WebSocket::makeHandler (desc_);
```

Comment:
Since we move from `handler` let's move it inside the scope (but before the lock).